lib/c-api/build.rs

```diff
@@ -40,67 +25,356 @@ fn main() {
+                r#"
+// The `{feature}` feature has been enabled for this build.
+#define {define}
+"#,
+                feature = $feature,
+                define = $c_define,
+            ));
+        }
+    };
+}
+
+fn main() {
+    let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+    let out_dir = env::var("OUT_DIR").unwrap();
+
+    build_wasm_c_api_headers(&crate_dir, &out_dir);
+    build_wasmer_headers(&crate_dir, &out_dir);
+}
+
+fn build_wasm_c_api_headers(crate_dir: &str, out_dir: &str) {
+    let mut crate_header_file = PathBuf::from(crate_dir);
+    crate_header_file.push("wasmer_wasm");
+
+    let mut out_header_file = PathBuf::from(out_dir);
+    out_header_file.push("wasmer_wasm");
+
+    let mut pre_header = format!(
+        r#"// The Wasmer C/C++ header file compatible with the `wasm-c-api` standard API.
+
+#if !defined(WASMER_WASM_H_MACROS)
+
+#define WASMER_WASM_H_MACROS
+{pre_header}"#,
+        pre_header = PRE_HEADER
+    );
+
+    map_feature_as_c_define!("jit", JIT_FEATURE_AS_C_DEFINE, pre_header);
+    map_feature_as_c_define!("compiler", COMPILER_FEATURE_AS_C_DEFINE, pre_header);
+    map_feature_as_c_define!("wasi", WASI_FEATURE_AS_C_DEFINE, pre_header);
+    map_feature_as_c_define!("emscripten", EMSCRIPTEN_FEATURE_AS_C_DEFINE, pre_header);
+    map_feature_as_c_define!(
+        "include-deprecated",
+        INCLUDE_DEPRECATED_FEATURE_AS_C_DEFINE,
+        pre_header
+    );
+
+    // Close pre header.
+    pre_header.push_str(
+        r#"
+#endif // WASMER_WASM_H_MACROS
 
-"#
-    .to_string();
 
-    #[cfg(feature = "wasi")]
+//
+// OK, here we go. The code below is automatically generated.
+//
+"#,
+    );
+
+    // C bindings.
     {
-        pre_header += "#define WASMER_WASI_ENABLED\n";
+        // Generate the bindings in the `OUT_DIR`.
+        out_header_file.set_extension("h");
+
+        #[allow(unused_mut)]
+        let mut builder = Builder::new()
+            .with_language(Language::C)
+            .with_crate(crate_dir)
+            .with_include_guard("WASMER_WASM_H")
+            .with_header(&pre_header)
+            .with_define("target_family", "windows", "_WIN32")
+            .with_define("target_arch", "x86_64", "ARCH_X86_64")
+            .with_define("feature", "jit", JIT_FEATURE_AS_C_DEFINE)
+            .with_define("feature", "compiler", COMPILER_FEATURE_AS_C_DEFINE)
+            .with_define("feature", "wasi", WASI_FEATURE_AS_C_DEFINE)
+            .with_define("feature", "emscripten", EMSCRIPTEN_FEATURE_AS_C_DEFINE)
+            .with_define(
+                "feature",
+                "include-deprecated",
+                INCLUDE_DEPRECATED_FEATURE_AS_C_DEFINE,
+            )
+            // Do not include `wasm.h` as it doesn't contain any
+            // documentation.
+            // We export the same functions, with documentations and
+            // examples, which is better for the user.
+            //
+            //.with_include("wasm.h")
```

Comment:
I think we should generally use the standard `wasm.h` in all our tests to make sure we're conforming to the signatures in it.  Alternatively we can have some kind of verification step where we diff our header file against it.  I'm worried that distributing a custom version of wasm.h is a good way to accidentally not support the real wasm C API

Reply:
I agree with that.

Reply:
`wasm.h` is now included, and functions and types are not exported from our code any longer by cbindgen.